source/opt/types.cpp

```diff
@@ -383,17 +383,46 @@ void SampledImage::GetExtraHashWords(
   image_type_->GetHashWords(words, seen);
 }
 
-Array::Array(Type* type, uint32_t length_id)
-    : Type(kArray), element_type_(type), length_id_(length_id) {
+Array::Array(Type* type, uint32_t length_id, uint32_t length_spec_id)
+    : Type(kArray),
+      element_type_(type),
+      length_id_(length_id),
+      length_spec_id_(length_spec_id),
+      length_constant_type_(nullptr),
+      length_constant_words_() {
   assert(!type->AsVoid());
+  assert(length_spec_id != 0u);
+}
+
+Array::Array(Type* type, uint32_t length_id, const Type* length_constant_type,
+             Operand::OperandData length_constant_words)
+    : Type(kArray),
+      element_type_(type),
+      length_id_(length_id),
+      length_spec_id_(0u),
+      length_constant_type_(length_constant_type),
+      length_constant_words_(length_constant_words) {
+  assert(!type->AsVoid());
+  assert(length_constant_type && length_constant_type->AsInteger());
 }
 
 bool Array::IsSameImpl(const Type* that, IsSameCache* seen) const {
   const Array* at = that->AsArray();
   if (!at) return false;
-  return length_id_ == at->length_id_ &&
-         element_type_->IsSameImpl(at->element_type_, seen) &&
-         HasSameDecorations(that);
+  bool is_same = element_type_->IsSameImpl(at->element_type_, seen) &&
+                 HasSameDecorations(that);
+  // If it is a specialized constant
+  if (length_spec_id_ != 0u) {
+    // ensure they have the same SpecId
+    is_same = is_same && length_spec_id_ == at->length_spec_id_;
+  } else {
+    // else, ensure they have the same length literal number.
+    is_same =
+        is_same &&
+        length_constant_type_->IsSameImpl(at->length_constant_type_, seen) &&
+        length_constant_words_ == at->length_constant_words_;
```

Comment:
This seems to be the only place where these two new members are used.  In general, the constant manager will be available right?  Couldn't you use the constant manager instead?  Say the Array class hold a pointer to a constant and this check would just make sure they are the same constant.

Reply:
AFAIK using the constant manager here will cause some issues. The constant manager uses the type manager to unify Constants, so using it in here (should) cause an infinite loop between `IRContext::BuildTypeManager` and `BuildConstantManager`. On the other hand, valid SPIR-V should present types and constants in a forward-only order, so this might not be an issue if instruction parsing is interleaved between the two managers (requiring a more significant refactor).

Reply:
Ideally yes, but we might have to create the constant object manually rather than going through the constant manager, since `IntConstant` requires an `Integer` object, and then `Array` would require an `IntConstant`. So we would need to analyse some of the types to construct some of the types object, that can then be used to create the constant objects, which in turn will be used to create more types object, which will allow creating more constant objects…